hashintel/hash
# Load entities into blocks fixes

## 🌟 What is the purpose of this PR?



This is a follow up to #480 and fixes an issue with loading entities into blocks. There was a temporary [hack](https://github.com/hashintel/hash/blob/99f1d29dee0382dcd7a9fbee28b444bba9457425/packages/hash/shared/src/entityStorePlugin.ts#L206-L209) added in #480 that prevents blocks from being reset to their initial state once the block's data has been swapped. This PR removes the need for that workaround. 

## 🔗 Related links





- [Fix bug with loading entities into blocks](https://app.asana.com/0/1202055186269230/1202106892392942) _(internal)_

## 🔍 What does this change?




- Adds a new `UpdatePageAction` => SwapBlockData  which handles when a block's data has been swapped with a different entity.
- Adds an `updateBlockData` method to block model to handle updating a block's data. This also includes deleting the existing link and creating a new link if the new block data is a different entity than the existing one.
- Updates `updatePageContents` to handle `SwapBlockData` action. This involves deleting the link between a block and it's existing data and creating a new link between the block and the new entity
- Filters the entities displayed in LoadEntityMenu to remove duplicate entities and the current entity the block is linked to


## 📜 Does this require a change to the docs?



- No

## ⚠️ Known issues


- This doesn't fix the reset bug that happens when a block's data is updated via the block config menu
- Loading an entity into an empty draft block doesn't work.. See [comment](https://github.com/hashintel/hash/pull/506/files#discussion_r867602284)
- Bug where a variant block (like the Youtube block) gets reset to the default block (embed block) still happens... See [comment](https://github.com/hashintel/hash/pull/506/files#discussion_r867602284)
- Loading entities into text block doesn't work. Should be handled in this [task](https://app.asana.com/0/1202176479837532/1202128324698426)


## 🐾 Next steps


- It will be a good idea to work on [Refactor all uses of the legacy `__linkedData` field](https://app.asana.com/0/1202134947335476/1201850801682936) before addressing the Known issues

## 🛡 What tests cover this?



- Existing tests

## ❓ How to test this?



1.  Create multiple code blocks (or any other block) and put some data in it. This helps generate a list of entities to switch between
1.  Try swapping the entity of any of the blocks via the block context menu. If the swap entity option doesn't show up in the block context menu, add `hash-load-entity-ui` to local storage
1.  Confirm swapping works

## 📹 Demo



https://user-images.githubusercontent.com/16928688/167330063-1cb56ecf-f8ab-4aac-9ee6-db998542607b.mp4
